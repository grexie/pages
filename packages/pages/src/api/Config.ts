import { ResourceMetadata } from './Resource.js';

export interface Config extends Record<string, any> {
  render: boolean;
  metadata: ResourceMetadata;
}
