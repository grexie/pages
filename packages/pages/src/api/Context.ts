export interface ContextOptions {}

export class Context {
  readonly isRuntime: boolean;
  readonly isServer: boolean;
  readonly isBuild: boolean;

  constructor({
    isServer = false,
    isBuild = false,
  }: ContextOptions & { isServer?: boolean; isBuild?: boolean }) {
    this.isRuntime = !isServer && !isBuild;
    this.isServer = isServer;
    this.isBuild = isBuild;
  }
}
