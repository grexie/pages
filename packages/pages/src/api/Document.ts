import { ReactElement } from 'react';
import { ResourceContext } from '../hooks';
import { Resource } from './Resource';

export interface DocumentOptions {
  resource: Resource;
  resourceContext: ResourceContext;
  initialProps?: DocumentProps;
}

export interface DocumentProps {
  title?: string;
  scripts?: string | string[];
  children: ReactElement[];
}

export class Document {
  readonly resourceContext: ResourceContext;
  readonly resource: Resource;
  readonly props: DocumentProps = { children: [] };

  constructor({
    initialProps = { children: [] },
    resource,
    resourceContext,
  }: DocumentOptions) {
    this.resourceContext = resourceContext;
    this.resource = resource;
    let { title, scripts, children } = initialProps;

    Object.assign(this.props, {
      title,
      scripts,
      children: [...this.props.children, ...children],
    });
  }
}
