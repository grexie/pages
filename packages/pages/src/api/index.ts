export * from './Context';
export * from './Provider';
export * from './Registry';
export * from './Handler';
export * from './Resource';
export * from './Source';
