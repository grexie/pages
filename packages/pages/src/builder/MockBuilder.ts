import { BuildContext } from './BuildContext';
import { Provider, ResourceMetadata } from '../api';
import { vol, Volume } from 'memfs';
import { FileSystem, WritableFileSystem } from '@grexie/builder';
import path from 'path';
import { Config } from './ConfigContext';
import YAML from 'yaml';
import fs from 'fs';

export class JSXSource {
  readonly builder: MockBuilder;
  readonly filename: string;
  readonly jsx: string;
  readonly imports: { importName: string; name: string; path: string }[] = [];
  readonly metadata: ResourceMetadata;
  readonly _code: string[] = [];
  readonly decorators: string[] = [];

  constructor(
    builder: MockBuilder,
    filename: string,
    jsx: string,
    metadata: ResourceMetadata
  ) {
    this.builder = builder;
    this.filename = filename;
    this.jsx = jsx;
    this.metadata = metadata;
  }

  code(code: string) {
    this._code.push(code);
    return this;
  }

  decorate(decorator: string) {
    this.decorators.push(decorator);
    return this;
  }

  importDefault(path: string, name: string) {
    this.imports.push({ importName: 'default', name, path });
    return this;
  }

  import(path: string, importName: string, name: string = importName) {
    this.imports.push({ importName, name, path });
    return this;
  }

  write() {
    const decorate = (
      code: string,
      decorators: string[] = this.decorators.slice()
    ): string => {
      let decorator = decorators.pop();
      if (!decorator) {
        return code;
      }

      return decorate(`${decorator}(${code})`, decorators);
    };
    const source = `
import React from 'react';
${this.imports
  .map(({ importName, name, path }) => {
    if (importName === 'default') {
      return `import ${name} from ${JSON.stringify(path)};`;
    } else if (importName === name) {
      return `import { ${name} } from ${JSON.stringify(path)};`;
    } else {
      return `import { ${importName} as ${name} } from ${JSON.stringify(
        path
      )};`;
    }
  })
  .join('\n')}

export default ${decorate(`({ children }) => {
  ${this._code.join('\n')}
  return (
    ${this.jsx}
  );
}`)};

export const resource = async (context) => {
  Object.assign(context.metadata, ${JSON.stringify(this.metadata)});
  return context.create();
}
`;
    this.builder.write(this.filename, source);
    return source;
  }
}

export interface MockBuilderOptions {
  builder?: MockBuilder;
  clean?: boolean;
}

export class MockBuilder extends BuildContext {
  constructor({ builder, clean = true }: MockBuilderOptions = {}) {
    const providers = [
      {
        provider: Provider,
      },
    ];
    let volume = builder?.defaultFiles ?? (new Volume() as WritableFileSystem);
    let rootDir = '/pages';

    if (process.env.PAGES_TEST_ROOT) {
      process.env.PAGES_CACHE = path.join(
        process.env.PAGES_TEST_ROOT,
        '.cache'
      );
      process.env.GREXIE_BUILDER_CACHE_HASH = 'false';
      rootDir = path.resolve(process.cwd(), process.env.PAGES_TEST_ROOT);
      volume = fs;
    }

    super({
      rootDir,
      fs: volume,
      defaultFiles: volume,
      providers,
    });
    this.defaultFiles = volume;
    if (clean) {
      volume.rmSync(this.rootDir, { recursive: true, force: true });
    }
    volume.mkdirSync(this.rootDir, { recursive: true });
    this.fs.add(path.resolve(this.pagesDir, '..', '..'), fs);
    this.fs.add(this.outputDir, volume, true);
  }

  create(options: Omit<MockBuilderOptions, 'builder'> = {}) {
    return new MockBuilder({ builder: this, ...options });
  }

  write(filename: string, data: Buffer | string) {
    this.builder.defaultFiles.mkdirSync(
      path.dirname(path.resolve(this.rootDir, filename)),
      { recursive: true }
    );
    return this.builder.defaultFiles.writeFileSync(
      path.resolve(this.rootDir, filename),
      data
    );
  }

  touchAll(dirname: string = this.rootDir) {
    this.builder.fs.readdirSync(dirname).forEach((file: any) => {
      const filename = path.resolve(
        dirname,
        file.name ? file.name : file.toString()
      );

      if ([this.cacheDir, this.outputDir].includes(filename)) {
        return;
      }

      const stats = this.builder.fs.statSync(filename);

      if (stats.isDirectory()) {
        this.touchAll(filename);
      } else {
        this.touch(filename);
      }
    });
  }

  touch(filename: string) {
    filename = path.resolve(this.rootDir, filename);
    this.builder.defaultFiles.writeFileSync(
      filename,
      this.builder.defaultFiles.readFileSync(filename)
    );
  }

  cleanOutput() {
    this.builder.fs.rmSync(this.outputDir, { recursive: true, force: true });
  }

  addSource(filename: string, jsx: string, metadata: ResourceMetadata = {}) {
    return new JSXSource(this, filename, jsx, metadata);
  }

  addConfig(filename: string, config: Config) {
    return this.write(filename, YAML.stringify(config));
  }

  output() {
    const out: { filename: string; data: string }[] = [];
    const stack: string[] = [this.outputDir];
    let el: string | undefined;
    while ((el = stack.shift())) {
      const files = this.fs.readdirSync(el);
      files.forEach((file: any) => {
        const filename = path.resolve(
          el!,
          file.name ? file.name : file.toString()
        );
        const stats = this.fs.statSync(filename);
        if (stats.isDirectory()) {
          stack.push(filename);
        } else {
          const data = this.fs.readFileSync(filename).toString();
          out.push({ filename: path.relative(this.outputDir, filename), data });
        }
      });
    }
    out.sort((a, b) => a.filename.length - b.filename.length);
    return out;
  }

  async build() {
    let sources = await this.registry.list();
    const stats = await this.builder.build(sources);

    if (stats.hasErrors()) {
      throw stats.compilation.errors;
    }
    return stats;
  }

  async watch() {
    let sources = await this.registry.list();
    return this.builder.watch(sources);
  }
}
