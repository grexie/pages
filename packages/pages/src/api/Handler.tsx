import { createElement, ComponentType } from 'react';
import { Resource, ResourceMetadata } from './Resource.js';
import { hydrateRoot } from 'react-dom/client';
import type { SourceContext } from '../builder/SourceContext.js';
import { compose } from '@grexie/compose';
import {
  ResourceContext,
  StylesContext,
  withContext,
  withDocument,
  withResource,
  withResourceContext,
  withStyles,
} from '../hooks/index.js';
import { withHydratedDocumentComponent } from '../components/Document.js';
import { Context } from './Context.js';
import { withLazy } from '../hooks/useLazy.js';

export interface Handler<
  P = any,
  M extends ResourceMetadata = any,
  R extends Resource<M> = Resource<M>
> {
  default?: ComponentType<P>;
  resource?: ((context: SourceContext) => Promise<R>) | Promise<R> | R;
}

export const wrapHandler = (
  resource: Resource,
  handler: Handler,
  ...composables: any[]
) => {
  return compose(...composables, withResource({ resource }), handler as any);
};

export const hydrate = (resource: Resource, handler: any) => {
  if (typeof window === 'undefined') {
    return;
  }

  const data = (window as any)['__PAGES_DATA__'];
  if (typeof data === 'undefined') {
    return;
  }

  if (data.slug !== resource.slug) {
    return;
  }

  const styles = new StylesContext();
  const context = new Context({});
  const resourceContext = new ResourceContext();

  const component = compose(
    withLazy,
    withContext({ context }),
    withStyles({ styles }),
    withResourceContext({ resourceContext }),
    withDocument({ resourceContext, resource }),
    withHydratedDocumentComponent,
    handler
  );

  const element = createElement(component as any);

  if ((window as any).__PAGES_ROOT__) {
    (window as any).__PAGES_ROOT__.render(element);
  } else {
    (window as any).__PAGES_ROOT__ = hydrateRoot(
      document.querySelector('#__pages_root')!,
      element
    );
  }
};
