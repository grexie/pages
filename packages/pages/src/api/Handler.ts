import { ComponentType } from 'react';
import { Resource, ResourceMetadata } from './Resource';
import { SourceContext } from '../builder/SourceContext';
import { compose } from '@grexie/compose';
import { withResource } from '../hooks';

export interface Handler<
  P = any,
  M extends ResourceMetadata = any,
  R extends Resource<M> = Resource<M>
> {
  default?: ComponentType<P>;
  resource?: ((context: SourceContext) => Promise<R>) | Promise<R> | R;
}

export const wrapHandler = (
  exports: any,
  resource: Resource,
  handler: Handler,
  ...composables: any[]
) => {
  if (typeof handler.default === 'function') {
    const Component = compose(
      ...composables,
      withResource({ resource }),
      handler.default as any
    );
    exports.default = Component;
    exports.resource = resource;
    if (!exports.__esModule) {
      exports.__esModule = true;
    }
  }
};
