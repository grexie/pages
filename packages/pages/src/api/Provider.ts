import glob from 'glob';
import _path from 'path';
import { promisify } from 'util';
import { Source } from './Source';
import { createResolver, ResolvablePromise } from '../utils/resolvable';
import { ProviderOptions, ListOptions } from './Registry';
import { BuildContext } from '../builder/BuildContext';

const globAsync = promisify(glob);

export class Provider {
  readonly context: BuildContext;
  #sources: ResolvablePromise<Record<string, Source>>;

  constructor({ context }: ProviderOptions) {
    this.context = context;

    this.#sources = createResolver();
    this.scan();
  }

  private async create(
    filename: string,
    rootDir: string
  ): Promise<Source | undefined> {
    const path = this.context.builder.filenameToPath(filename, rootDir);

    return new Source({
      filename,
      path,
    });
  }

  private async scan(): Promise<void> {
    const files = await globAsync('**/*', {
      cwd: this.context.rootDir,
      nodir: true,
      dot: true,
      ignore: [
        _path.relative(
          this.context.rootDir,
          _path.join(this.context.outputDir, '**')
        ),
        _path.relative(
          this.context.rootDir,
          _path.join(this.context.cacheDir, '**')
        ),
        'node_modules/**',
        '.git/**',
        '.github/**',
        '.gitignore',
        'yarn-error.log',
        'yarn.lock',
        'package-lock.json',
      ],
    });

    this.#sources.resolve(
      (
        (
          await Promise.all(
            files.map(async filename =>
              this.create(
                _path.resolve(this.context.rootDir, filename),
                process.cwd()
              )
            )
          )
        ).filter(source => !!source) as Source[]
      ).reduce(
        (resources, resource) => ({
          ...resources,
          [resource.path.join('/')]: resource,
        }),
        {}
      )
    );
  }

  async list({ type, path, slug }: ListOptions = {}): Promise<Source[]> {
    let sourcesMap = await this.#sources;
    let sources = Object.values(sourcesMap);

    // if (typeof type !== 'undefined') {
    //   if (typeof type === 'string') {
    //     type = [type];
    //   }
    //   const typeMap: Record<string, boolean> = type.reduce(
    //     (a, b) => ({ ...a, [b]: true }),
    //     {}
    //   );
    //   resources = (
    //     await Promise.all(
    //       resources.map(async resource => ({
    //         resource,
    //         metadata: await resource.metadata(),
    //       }))
    //     )
    //   )
    //     .filter(({ metadata }) => typeMap[metadata.type ?? ''])
    //     .map(({ resource }) => resource);
    // }
    if (typeof path !== 'undefined') {
      if (typeof path[0] === 'string') {
        path = [path as string[]];
      }
      if (path.length === 0) {
        path = [[]];
      }
      slug = path.map(path => (path as string[]).join('/'));
    }
    if (typeof slug !== 'undefined') {
      if (typeof slug === 'string') {
        slug = [slug];
      }

      if (slug.length === 1) {
        const resource = sourcesMap[slug[0]];
        sources = resource ? [resource] : [];
      } else {
        const slugMap: Record<string, boolean> = slug.reduce(
          (a, b) => ({ ...a, [b]: true }),
          {}
        );
        sources = sources.filter(source => slugMap[source.slug]);
      }
    }
    return sources;
  }

  async has(path: string[]): Promise<boolean> {
    const filepath = path.join('/');
    const resources = await this.#sources;
    return filepath in resources;
  }
}
