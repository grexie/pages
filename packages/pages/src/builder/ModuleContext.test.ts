import path from 'path';
import webpack from 'webpack';
import { Source } from '../api';
import { MockBuilder } from './MockBuilder';

const createFixtures = (builder: MockBuilder) => {
  builder.addConfig('.pages.yml', {
    metadata: {
      layout: ['./layout'],
    },
  });
  builder
    .addSource(
      'layout.jsx',
      `<><Head><title>{title} | Test Site</title></Head><h1>{title}</h1>{children}</>`,
      {
        layout: null,
      }
    )
    .code('const { title } = useMetadata();')
    .decorate('withOnce')
    .import('@grexie/pages', 'withOnce')
    .import('@grexie/pages', 'Head')
    .import('@grexie/pages', 'useMetadata')
    .write();
  builder
    .addSource('test1.jsx', `<>Test1<Test2/></>`, { title: 'Test 1' })
    .importDefault('./test2', 'Test2')
    .write();
  builder.addSource('test2.jsx', `<>Test2</>`, { title: 'Test 2' }).write();
  builder
    .addSource('test3.jsx', `<>Test3</>`, { title: 'Test 3', layout: null })
    .write();
};

describe('ModuleContext', () => {
  let builder: MockBuilder;

  beforeEach(async () => {
    builder = new MockBuilder();
  });

  describe('build cache', () => {
    beforeEach(async () => {
      createFixtures(builder);
    });

    it('should enumerate all sources', async () => {
      const sources = await builder.registry.list();
      const configs = await builder.registry.listConfig();
      const serializeSource = (source: Source) => {
        const { path, slug } = source;
        return { path, slug };
      };
      expect(sources.map(serializeSource)).toMatchSnapshot();
      expect(configs.map(serializeSource)).toMatchSnapshot();
    });

    it('should build once', async () => {
      await builder.build();
      expect(builder.output()).toMatchSnapshot();
    });

    it('should build multiple times', async () => {
      await builder.build();
      const output = builder.output();
      expect(output).toMatchSnapshot();

      const builder2 = builder.create({ clean: false });
      builder2.cleanOutput();
      builder2.touchAll();
      await builder2.build();
      expect(builder2.output()).toEqual(output);

      const builder3 = builder.create({ clean: false });
      builder3.cleanOutput();
      builder3.touchAll();
      await builder3.build();
      expect(builder3.output()).toEqual(output);

      const builder4 = builder.create({ clean: false });
      builder4.cleanOutput();
      builder4.touchAll();
      await builder4.build();
      expect(builder4.output()).toEqual(output);
    }, 20000);

    it('should only rebuild files which have changed', async () => {
      await builder.build();

      const builder2 = builder.create({ clean: false });
      builder2.touch('test1.jsx');
      builder2.cleanOutput();
      await builder2.build();
      expect(builder2.output()).toMatchSnapshot();
    });

    it('should rebuild dependent resources', async () => {
      await builder.build();

      const builder2 = builder.create({ clean: false });
      builder2.touch('test2.jsx');
      builder2.cleanOutput();
      await builder2.build();
      expect(builder2.output()).toMatchSnapshot();
    });

    it('should rebuild dependent resources from a layout change', async () => {
      await builder.build();

      const builder2 = builder.create({ clean: false });
      builder2.touch('layout.jsx');
      builder2.cleanOutput();
      await builder2.build();
      expect(builder2.output()).toMatchSnapshot();
    });

    it('should be the same source on recompile', async () => {
      const builder1 = new MockBuilder();
      builder1.addSource('index.jsx', `<>Home</>`, { title: 'Home' }).write();
      const stats1 = await builder1.build();
      const factory1 = builder1.modules.createModuleFactory(stats1.compilation);
      const module1 = await builder1.modules.require(
        factory1,
        builder1.rootDir,
        path.resolve(builder1.rootDir, 'index.jsx')
      );

      const builder2 = builder1.create({ clean: false });
      const sources = await builder2.registry.list();
      const config = await builder2.builder.config(sources);
      const compiler = webpack(config);
      compiler.inputFileSystem = builder2.fs;
      compiler.outputFileSystem = builder2.fs;
      const compilation = compiler.newCompilation(
        compiler.newCompilationParams()
      );
      const factory2 = builder2.modules.createModuleFactory(compilation);
      const module2 = await builder2.modules.require(
        factory2,
        builder2.rootDir,
        path.resolve(builder2.rootDir, 'index.jsx')
      );

      expect(module2.source).toEqual(module1.source);
    });

    it('should create after a build and evict', async () => {
      const builder = new MockBuilder();
      const filename = path.resolve(builder.rootDir, 'index.jsx');

      const source = builder
        .addSource('index.jsx', `<>Home</>`, { title: 'Home' })
        .write();
      const sources = await builder.registry.list();
      const config = await builder.builder.config(sources);

      const compiler = webpack(config);
      compiler.inputFileSystem = builder.fs;
      compiler.outputFileSystem = builder.fs;
      const compilation = compiler.newCompilation(
        compiler.newCompilationParams()
      );

      const factory1 = builder.modules.createModuleFactory(compilation);

      const module1 = await builder.modules.require(
        factory1,
        builder.rootDir,
        filename
      );
      module1.load(module);

      await builder.modules.evict(factory1, filename, {
        recompile: true,
      });

      const factory2 = builder.modules.createModuleFactory(compilation);
      const compiled = await builder.modules.compiler.compile({
        source,
        filename,
        presets: [require('@babel/preset-react')],
      });
      const module2 = await builder.modules.create(
        factory2,
        new webpack.Module(module1.webpackModule.type),
        filename,
        compiled.source
      );

      expect(module2.source.replace(/\s+/g, ' ')).not.toEqual(
        module1.source.replace(/\s+/g, ' ')
      );
    });

    it('should rebuild from config change', async () => {
      await builder.build();

      const builder2 = builder.create({ clean: false });
      builder2.touch('.pages.yml');
      builder2.cleanOutput();
      await builder2.build();
      expect(builder2.output()).toMatchSnapshot();
    }, 20000);
  });
});
