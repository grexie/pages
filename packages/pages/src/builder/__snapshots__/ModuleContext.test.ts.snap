// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleContext build cache should build multiple times 1`] = `
Array [
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 1 | Test Site</title></head><body><h1>Test 1</h1>Test1<!-- -->Test2<!-- --></body></html>",
    "filename": "test1/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 2 | Test Site</title></head><body><h1>Test 2</h1>Test2<!-- --></body></html>",
    "filename": "test2/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/></head><body>Test3<!-- --></body></html>",
    "filename": "test3/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title> | Test Site</title></head><body><h1></h1></body></html>",
    "filename": "layout/index.html",
  },
]
`;

exports[`ModuleContext build cache should build once 1`] = `
Array [
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 1 | Test Site</title></head><body><h1>Test 1</h1>Test1<!-- -->Test2<!-- --></body></html>",
    "filename": "test1/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 2 | Test Site</title></head><body><h1>Test 2</h1>Test2<!-- --></body></html>",
    "filename": "test2/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/></head><body>Test3<!-- --></body></html>",
    "filename": "test3/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title> | Test Site</title></head><body><h1></h1></body></html>",
    "filename": "layout/index.html",
  },
]
`;

exports[`ModuleContext build cache should enumerate all sources 1`] = `
Array [
  Object {
    "path": Array [
      "layout",
    ],
    "slug": "layout",
  },
  Object {
    "path": Array [
      "test1",
    ],
    "slug": "test1",
  },
  Object {
    "path": Array [
      "test2",
    ],
    "slug": "test2",
  },
  Object {
    "path": Array [
      "test3",
    ],
    "slug": "test3",
  },
]
`;

exports[`ModuleContext build cache should enumerate all sources 2`] = `
Array [
  Object {
    "path": Array [],
    "slug": "",
  },
  Object {
    "path": Array [],
    "slug": "",
  },
]
`;

exports[`ModuleContext build cache should only rebuild files which have changed 1`] = `
Array [
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 1 | Test Site</title></head><body><h1>Test 1</h1>Test1<!-- -->Test2<!-- --></body></html>",
    "filename": "test1/index.html",
  },
]
`;

exports[`ModuleContext build cache should rebuild dependent resources 1`] = `
Array [
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 1 | Test Site</title></head><body><h1>Test 1</h1>Test1<!-- -->Test2<!-- --></body></html>",
    "filename": "test1/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 2 | Test Site</title></head><body><h1>Test 2</h1>Test2<!-- --></body></html>",
    "filename": "test2/index.html",
  },
]
`;

exports[`ModuleContext build cache should rebuild dependent resources from a layout change 1`] = `
Array [
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 1 | Test Site</title></head><body><h1>Test 1</h1>Test1<!-- -->Test2<!-- --></body></html>",
    "filename": "test1/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 2 | Test Site</title></head><body><h1>Test 2</h1>Test2<!-- --></body></html>",
    "filename": "test2/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title> | Test Site</title></head><body><h1></h1></body></html>",
    "filename": "layout/index.html",
  },
]
`;

exports[`ModuleContext build cache should rebuild from config change 1`] = `
Array [
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 1 | Test Site</title></head><body><h1>Test 1</h1>Test1<!-- -->Test2<!-- --></body></html>",
    "filename": "test1/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title>Test 2 | Test Site</title></head><body><h1>Test 2</h1>Test2<!-- --></body></html>",
    "filename": "test2/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/></head><body>Test3<!-- --></body></html>",
    "filename": "test3/index.html",
  },
  Object {
    "data": "<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><title> | Test Site</title></head><body><h1></h1></body></html>",
    "filename": "layout/index.html",
  },
]
`;
